proposals/handler.go

```diff
@@ -180,27 +179,53 @@ func (h *Handler) HandleSyncedBallot(ctx context.Context, expHash types.Hash32,
 	return nil
 }
 
-// collectHashes gathers all hashes in a proposal or ballot.
-func collectHashes(a any) []types.Hash32 {
-	p, ok := a.(types.Proposal)
-	if ok {
-		hashes := collectHashes(p.Ballot)
-		return append(hashes, types.TransactionIDsToHashes(p.TxIDs)...)
+func (h *Handler) HandleActiveSet(ctx context.Context, id types.Hash32, peer p2p.Peer, data []byte) error {
+	var set types.EpochActiveSet
+	if err := codec.Decode(data, &set); err != nil {
+		return fmt.Errorf("%w: malformed active set %s", pubsub.ErrValidationReject, err.Error())
 	}
+	h.fetcher.RegisterPeerHashes(peer, types.ATXIDsToHashes(set.Set))
+	return h.handleSet(ctx, id, set)
+}
 
-	b, ok := a.(types.Ballot)
-	if ok {
-		hashes := []types.Hash32{b.Votes.Base.AsHash32()}
-		if b.RefBallot != types.EmptyBallotID {
-			hashes = append(hashes, b.RefBallot.AsHash32())
+func (h *Handler) handleSet(ctx context.Context, id types.Hash32, set types.EpochActiveSet) error {
```

Comment:
add metrics for how many different active set hash we've seen from the network for this epoch. this would help us understand whether we should improve with base active set + diff

Reply:
i don't want to maintain map for that. we can check growth by looking table size and number or records in table (second we don't have, but should be added)